File: Algorithms/045.JumpGameII.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        int maxIndex = 0;
        int steps[n];
        for (int i=0; i < n; i++) {
            steps[i] = INT_MAX;
        }
        steps[0] = 0;
        for (int i=0; i < n && i <= maxIndex; i++) {
            for (int j=maxIndex+1; j <= i+nums[i] && j < n; j++) {
                if (steps[i]+1 < steps[j]) {
                    steps[j] = steps[i] + 1;
                }
            }
            if (i+nums[i] < maxIndex) {
                continue;
            }
            maxIndex = i + nums[i];
            if (maxIndex >= n-1) {
                break;
            }
        }
        return steps[n-1];
    }
};

File: Algorithms/py/045.JumpGameII.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def jump(self, nums):
        maximum, i, steps = 0, 0, [None for j in range(len(nums))]
        steps[0] = 0
        while i < len(nums) and i <= maximum:
            for j in range(maximum+1, min(i+nums[i]+1, len(nums))):
                if not steps[j]:
                    steps[j] = steps[i] + 1
            maximum = max(i + nums[i], maximum)
            if maximum >= len(nums)-1:
                return steps[-1]
            i += 1

    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        steps = [n-1]*n
        steps[0] = 0
        for i in range(n):
            for j in range(i+1, min(i+nums[i]+1, n)):
                steps[j] = min(steps[j], steps[i] + 1)
        #print(steps)
        return steps[-1]
